### ISSUE TITLE
Check constraints and autoincrement

### ISSUE THREAD

**Issue by pgulutzan** [tarantool/tarantool#3669]
I execute these statements:
I have Ubuntu 18.04. I have Tarantool 2.0, pulled from source today.
I execute these statements:
```SQL
CREATE TABLE t5 (s1 INT PRIMARY KEY AUTOINCREMENT, s2 CHAR(1), CHECK (s1 < 1));
INSERT INTO t5 (s2) VALUES (NULL);
INSERT INTO t5 VALUES (NULL,NULL);
select * from t5;
```

The first INSERT fails, this was issue#2981 but it was solved.
The second INSERT succeeds, and s1 = 2.

**Comment by Gerold103** [comment]
@kyukhin I do not think it is a bug - the sequence here works correctly - it grows and never decreases. The same example can be rewritten like this: begin, insert null, rollback - no errors, but sequence can not be rolled back, so next insert will use next sequence value.

**Comment by pgulutzan** [comment]
I apologize for not also giving examples of what I thought was
Tarantool's normal behaviour. Try these two.
drop table t5;
CREATE TABLE t5 (s1 INT PRIMARY KEY AUTOINCREMENT, s2 int, CHECK (s1 < 1));
INSERT INTO t5 VALUES (NULL, NULL);
INSERT INTO t5 VALUES (NULL, NULL);
SELECT * FROM t5;
The first INSERT succeeds. The second INSERT succeeds, and s1 = 2.
Thus "INSERT INTO t5 VALUES (NULL, NULL);" is not equivalent to
"INSERT INTO t5 (s2) VALUES (NULL);".
+
drop table t5;
CREATE TABLE t5 (s1 INT PRIMARY KEY AUTOINCREMENT, s2 int, CHECK (s2 < 1));
INSERT INTO t5 VALUES (NULL, 1);
INSERT INTO t5 VALUES (NULL, 0);
SELECT * FROM t5;
The first INSERT fails.
The second INSERT succeeds, and s1 = 1.
Not 2 but 1.
So:
(a) It is okay to say that the non-standard autoincrement depends on
sequences which are unaffected by transaction commit|rollback. We have
not clearly said so, but the SQL manual does use the word "sequence" and
Mr Shpilevoy is clearly saying so in his comment now.
(b) It is okay to say that a sequence's value is incremented at the start of
a statement's execution and abortion of statements does not affect it.
We have not explicitly said so, but it is implied by Mr Shpilevoy's example,
where he suggests that the original statement-abortion example can be rewritten as
a transaction-rollback example.
(c) It is okay to say that Tarantool autoincrement is not like MySQL
auto_increment (because MySQL would return s1 = 1 for the first example),
and that Tarantool autoincrement is not like SQLite autoincrement (because
tarantool does not re-use numbers after insertion of 9223372036854775807).
(d) It is okay to say that CHECK clauses are sometimes evaluated before
the sequence value goes up, and that CHECK clauses are sometimes evaluated after
the sequence value goes up, depending on what appear to be trivial
syntax differences.
(e) But it is not okay to say nothing. So call it a documentation error,
but please clarify that (a) and (b) and (c) and (d) are really expected.

**Comment by pgulutzan** [comment]
With non-standard syntax you can do whatever you decide is necessary, but of course it would be better if CHECK clauses were always enforced. Ignore my remark about MySQL -- I had forgotten that MySQL doesn't enforce CHECK clauses. I think the immediate issue with (d) can be solved by explaining why the results differ.

**Comment by kshcherbatov** [comment]
This problem would be closed when CK CONSTRAINTS server integration would be completed #3691.
```
tarantool> \set language sql
---
- true
...

tarantool> CREATE TABLE t5 (s1 INT PRIMARY KEY AUTOINCREMENT, s2 int, CHECK (s1 < 1));
---
- row_count: 1
...

tarantool> INSERT INTO t5 VALUES (NULL, NULL);
---
- error: 'Failed to execute SQL statement: Check constraint failed ''CK_CONSTRAINT_1_T5'':
    s1 < 1'
...

tarantool> INSERT INTO t5 VALUES (NULL, NULL);
---
- error: 'Failed to execute SQL statement: Check constraint failed ''CK_CONSTRAINT_1_T5'':
    s1 < 1'
...

tarantool> SELECT * FROM t5;
---
- metadata:
  - name: S1
    type: integer
  - name: S2
    type: integer
  rows: []
...
```

**Comment by kshcherbatov** [comment]
Closed with #3691.
